include/circt/Dialect/FIRRTL/FIRRTLTypesImpl.td

```diff
@@ -38,54 +38,86 @@ def SIntImpl : FIRRTLImplType<"SInt",
                               [WidthQualifiedTypeTrait, FieldIDTypeInterface],
                               "::circt::firrtl::IntType"> {
   let summary = "A signed integer type, whose width may not be known.";
-  let parameters = (ins "int32_t":$widthOrSentinel);
+  let parameters = (ins "int32_t":$widthOrSentinel, "bool":$isConst);
   let builders = [
-    TypeBuilder<(ins "std::optional<int32_t>":$width)>,
+    TypeBuilder<(ins "std::optional<int32_t>":$width, CArg<"bool", "false">:$isConst)>,
     TypeBuilder<(ins)>,
   ];
   let genVerifyDecl = true;
   let extraClassDeclaration = [{
     using WidthQualifiedTypeTrait<SIntType>::getWidth;
     using WidthQualifiedTypeTrait<SIntType>::hasWidth;
+    bool isConst() { return getIsConst(); }
   }];
 }
 
 def UIntImpl : FIRRTLImplType<"UInt",
                               [WidthQualifiedTypeTrait, FieldIDTypeInterface],
                               "::circt::firrtl::IntType"> {
   let summary = "An unsigned integer type, whose width may not be known.";
-  let parameters = (ins "int32_t":$widthOrSentinel);
+  let parameters = (ins "int32_t":$widthOrSentinel, "bool":$isConst);
   let builders = [
-    TypeBuilder<(ins "std::optional<int32_t>":$width)>,
+    TypeBuilder<(ins "std::optional<int32_t>":$width, CArg<"bool", "false">:$isConst)>,
     TypeBuilder<(ins)>,
   ];
   let genVerifyDecl = true;
   let extraClassDeclaration = [{
     using WidthQualifiedTypeTrait<UIntType>::getWidth;
     using WidthQualifiedTypeTrait<UIntType>::hasWidth;
+    bool isConst() { return getIsConst(); }
   }];
 }
 
 def ClockTypeImpl : FIRRTLImplType<"Clock", [FieldIDTypeInterface]> {
   let summary = "Clock signal";
+  let parameters = (ins "bool":$isConst);
+  let builders = [
+    TypeBuilder<(ins), [{
+      return $_get($_ctxt, false);
+    }]>
+  ];
+  let extraClassDeclaration = [{
+    bool isConst() { return getIsConst(); }
```

Comment:
I don't think you need all of these.  The storage should be doable with the FIRRTL type storage (like the recursive type properties).

Reply:
These functions in particular are just nicer accessors to `getIsConst`, which is generated from the `isConst` type parameter. I'm not sure if there's a way to directly generate the nicer name.

But yeah, it would be nicer to only have to declare it once on the base type. Not sure what you mean by recursive type properties though, that's just a function that dispatches to concrete subclasses.

Reply:
These are all removed now, with `isConst` being stored in `FIRRTLBaseTypeStorage`